analyze_impact.py

```diff
@@ -36,72 +45,105 @@ def get_linked_columns(
+    lineage_data = execute_stored_query(
+        QUERIES / 'lineage_analyze_impact.gql',
+        primaryUid=table_uid,
+        depthDownstream=1000,
+        depthUpstream=0,
+        popularity=[0, 4],
+        biLastUsedDays=90,
+        allowedList=column_uids,
+    )
 
-    args = parser.parse_args()
+    connected_columns = set()
 
-    # fast and dirty, doesn't handle quotes
-    table_path = tuple(args.full_table_name.split('.'))
+    edges = [
+        (source, destination)
+        for edge in lineage_data['lineage']['edges']
+        if (
+            # Filter out "off-chart" edges
+            (source := edge['sourceUid'])
+            and (destination := edge['destinationUid'])
+        )
+    ]
+
+    if column_uids:
+        graph = networkx.Graph()
+        graph.add_edges_from(edges)
+
+        connected_columns = {
+            connected_uid
+            for allowed_column in column_uids
+            for edge in networkx.dfs_edges(graph, source=allowed_column)
+            for connected_uid in edge
+        }
+
+    else:
+        connected_columns = {
+            connected_uid
+            for edge in edges
+            for connected_uid in edge
+        }
+
+    for tabular_entity in lineage_data['lineage']['entities']:
+        name = tabular_entity.get(
+            'prop', {},
+        ).get('path') or tabular_entity.get('name') or '???'
+        entity_type = tabular_entity['__typename']
+
+        rich.print(f'[purple]{entity_type}[/purple] [green]{name}[/green]')
+
+        for col in tabular_entity.get('columns', []):
+            uid = col.get('uid')
+            if uid is None:
+                continue  # that's not a column
+
+            if uid not in connected_columns:
+                continue
+
+            column_name = col.get('prop', {}).get('name') or col.get('name') or uid
+            rich.print(f'  - {column_name}')
+            if tags := col.get('tags', []):
+                console.print(f'    [i]Tags:[/i] ', end='')
+                for tag in tags:
+                    color = tag['color'].strip('#')
+                    console.print(
+                        tag['name'],
+                        style=f'#{color}', end='',
+                    )
+                    console.print()
+
+
+def analyze_impact(
+    data_source_id: Annotated[int, Argument(help='Data source ID.')],
+    table_path: Annotated[str, Argument(
+        help='Full table path, case sensitive. Format: DB.SCHEMA.TABLE',
+    )],
+    columns: Annotated[str, Option(help=(
+        'Print downstreams only for these columns of the table in question. '
+        'Format: `ID,COL1,ANOTHER_COL`. Case sensitive.'
+    ))] = '',
+):
+    """
+    Print downstream dependencies of a given table.
+
+    How to obtain Data Source ID:
+
+    * Open https://app.datafold.com,
+
+    * Go to **Settings** → Data Sources,
+
+    * Find your Data Source; the leftmost column in the Data Sources list is ID.
+    """
```

Comment:
If I'd read this, the first thing I'd do is check that the data_sources are secured by trying a random data_source ID that's not part of the same org 😅 

I guess we do some cross-org checks for fetching this data, right?

Reply:
Yes, that is a part of the backend.